arwen/contexts/runtime.go

```diff
@@ -833,6 +848,74 @@ func (context *runtimeContext) GetAllErrors() error {
+	context.currentGasTracer.traceGasEnabled = false
+}
+
+// TraceGasUsed adds the usedGas passed to the traced gas map with the key as the passed functionName
+func (context *runtimeContext) TraceGasUsed(functionName string, usedGas uint64) {
+	if context.currentGasTracer.traceGasEnabled && context.scAddress != nil {
+		scAddress := string(context.scAddress)
+		context.createGasTraceIfNil(scAddress, functionName)
+		context.currentGasTracer.gasTrace[scAddress][functionName] = append(context.currentGasTracer.gasTrace[scAddress][functionName], usedGas)
+	}
+}
+
+// UseAndTraceGas uses the passed gasToUse and traces it
+func (context *runtimeContext) UseAndTraceGas(functionName string, gasToUse uint64) {
+	metering := context.host.Metering()
+	metering.UseGas(gasToUse)
+	if context.currentGasTracer.traceGasEnabled && context.scAddress != nil {
+		scAddress := string(context.scAddress)
+		context.createGasTraceIfNil(scAddress, functionName)
+		context.currentGasTracer.gasTrace[scAddress][functionName] = append(context.currentGasTracer.gasTrace[scAddress][functionName], gasToUse)
+	}
+}
+
+// SetInitialGasInGasTrace sets the initial gas in the gasTrace map
+func (context *runtimeContext) SetInitialGasInGasTrace(functionName string) {
+	if context.currentGasTracer.traceGasEnabled && context.currentGasTracer.currentFunctionNameTraced == "" && context.scAddress != nil {
+		context.currentGasTracer.currentFunctionNameTraced = functionName
+		scAddress := string(context.scAddress)
+		context.createGasTraceIfNil(scAddress, functionName)
+		context.currentGasTracer.gasTrace[scAddress][functionName] = append(context.currentGasTracer.gasTrace[scAddress][functionName], context.host.Metering().GasLeft())
+	}
+}
+
+func (context *runtimeContext) createGasTraceIfNil(scAddress string, functionName string) {
+	if context.currentGasTracer.gasTrace[scAddress] == nil {
+		context.currentGasTracer.gasTrace[scAddress] = make(map[string][]uint64)
+	}
+	if context.currentGasTracer.gasTrace[scAddress][functionName] == nil {
+		context.currentGasTracer.gasTrace[scAddress][functionName] = make([]uint64, 0)
+	}
+}
+
+// ComputeAndSetUsedGasInGasTrace computes the used gas in the gasTrace map. This function should be used afte initial gas has been set in the gasTrace map using SetInitialGasInGasTrace.
+func (context *runtimeContext) ComputeAndSetUsedGasInGasTrace() {
+	if context.currentGasTracer.traceGasEnabled && context.scAddress != nil {
+		functionName := context.currentGasTracer.currentFunctionNameTraced
+		numberOfCalls := len(context.currentGasTracer.gasTrace[string(context.scAddress)][functionName])
+
+		gasLeft := context.host.Metering().GasLeft()
+		usedGas := context.currentGasTracer.gasTrace[string(context.scAddress)][functionName][numberOfCalls-1] - gasLeft
+		context.currentGasTracer.gasTrace[string(context.scAddress)][functionName][numberOfCalls-1] = usedGas
+
+		context.currentGasTracer.currentFunctionNameTraced = ""
```

Comment:
why do you rewrite this ?

Reply:
because if there is another API function called which uses `SetInitialGasInGasTrace()` and we don't rewrite `""`, then it would be able to trace gas

Reply:
We should not use construction like this. Function SetInitialGasInGasTrace should not exist